flocker/control/_persistence.py

```diff
@@ -21,16 +21,94 @@
 from twisted.internet.defer import succeed
 from twisted.internet.task import LoopingCall
 
-from ._model import SERIALIZABLE_CLASSES, Deployment
+from ._model import SERIALIZABLE_CLASSES, Deployment, Configuration
 
 
 # Serialization marker storing the class name:
 _CLASS_MARKER = u"$__class__$"
 
+# The latest configuration version. Configuration versions are
+# always integers.
+_CONFIG_VERSION = 2
 
-class _ConfigurationEncoder(JSONEncoder):
+# Map of serializable class names to classes
+_CONFIG_CLASS_MAP = {cls.__name__: cls for cls in SERIALIZABLE_CLASSES}
+
+
+class ConfigurationMigrationError(Exception):
+    """
+    Error raised when a configuration migration is unable to take place
+    or complete successfully.
+    """
+
+
+def migrate_configuration(source_version, target_version,
+                          config, migration_class):
+    """
+    Migrate a persisted configuration from one version to another
+    in sequential upgrades, e.g. a source version of 1 and target
+    version of 3 will perform two upgrades, from version 1 to 2,
+    followed by 2 to 3.
+
+    Calls the correct ``ConfigurationMigration`` class methods for
+    the suppled source and target versions.
+
+    :param int source_version: The version to migrate from.
+    :param int target_version: The version to migrate to.
+    :param bytes config: The JSON-encoded source configuration.
+    :param class migration_class: The class containing the methods
+        that will be used for migration.
+
+    :return bytes: The updated JSON configuration after migration.
+    """
+    upgraded_config = config
+    current_version = source_version
+    for upgrade_version in range(source_version + 1, target_version + 1):
+        with _LOG_UPGRADE(configuration=upgraded_config,
+                          source_version=current_version,
+                          target_version=upgrade_version):
+            migration_method = (
+                u"configuration_v%d_v%d"
+                % (current_version, upgrade_version)
+            )
+            try:
+                migration = getattr(migration_class, migration_method)
+            except AttributeError:
+                message = (
+                    u"Unable to find a migration path for a version " +
+                    unicode(source_version) + u" to version " +
+                    unicode(target_version) + u" configuration. " +
+                    u"No migration method exists for v" +
+                    unicode(current_version) + u" to v" +
+                    unicode(upgrade_version) + u"."
+                )
+                raise ConfigurationMigrationError(message)
```

Comment:
I'd suggest that this should actually be an error much earlier in the process.  For example, it would be fairly useless to try to run code that could upgrade, say, 1 to 2 and 3 to 4, but not from 2 to 3.

Please file a follow-up for detecting situations like this earlier (for example, in the test suite).